### PR TITLE
It versioning

### PR DESCRIPTION
**Pull request by stingion**

### CODE REVIEW

**Comment by segabriel** [comment]
in `io.scalecube.configuration.repository.couchbase.CouchbaseRepository#readAll`
```
Flux.from(
                    RxReactiveStreams.toPublisher(
                        bucket.query(
                            ViewQuery.from("keys", "by_keys")
                                .key(tenant + DELIMITER + repository)))
```

returns a response with empty rows in failing test (`io.scalecube.configuration.scenario.ReadListScenario`)

**Comment by stingion** [comment]
> in `io.scalecube.configuration.repository.couchbase.CouchbaseRepository#readAll`
> 
> ```
> Flux.from(
>                     RxReactiveStreams.toPublisher(
>                         bucket.query(
>                             ViewQuery.from("keys", "by_keys")
>                                 .key(tenant + DELIMITER + repository)))
> ```
> 
> returns a response with empty rows in failing test (`io.scalecube.configuration.scenario.ReadListScenario`)

**Comment by stingion** [comment]
> in `io.scalecube.configuration.repository.couchbase.CouchbaseRepository#readAll`
> 
> ```
> Flux.from(
>                     RxReactiveStreams.toPublisher(
>                         bucket.query(
>                             ViewQuery.from("keys", "by_keys")
>                                 .key(tenant + DELIMITER + repository)))
> ```
> 
> returns a response with empty rows in failing test (`io.scalecube.configuration.scenario.ReadListScenario`)

I know this issue (I told about it on stand up); if you stop a debug there and then proceed it will be work fine and manually it work fine.

File: scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java [hunk 1]
```diff
@@ -31,74 +38,124 @@
...
+                        : version <= vl.size() ? vl.get(version - 1) : null)
+            .filter(v -> v != null)
+            .collect(Collectors.toSet());
+    return Flux.fromIterable(values);
   }
 
   @Override
   public Flux<HistoryDocument> readHistory(String tenant, String repository, String key) {
-    throw new NotImplementedException();
+    AtomicInteger version = new AtomicInteger(0);
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java:60]: What is it?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java [hunk 2]
```diff
@@ -31,74 +38,124 @@
...
-    return entries(new Repository(tenant, repository));
+    Set<Document> values =
+        getRepository(new Repository(tenant, repository)).values().stream()
+            .map(
+                vl ->
+                    version == null
+                        ? vl.get(vl.size() - 1)
+                        : version <= vl.size() ? vl.get(version - 1) : null)
+            .filter(v -> v != null)
+            .collect(Collectors.toSet());
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java:54]: Is it possible don't collect it again?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java [hunk 3]
```diff
@@ -31,74 +38,124 @@
...
+            .map(doc -> new HistoryDocument(version.incrementAndGet(), doc.value()))
+            .collect(Collectors.toSet());
+    return Flux.fromIterable(values);
   }
 
   @Override
-  public Mono<Document> save(String tenant, String repository, Document document) {
-    return put(new Repository(tenant, repository), document.key(), document);
+  public Mono<Document> save(String tenant, String repository, Document doc) {
+    Repository repo = new Repository(tenant, repository);
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java]: Please wrap it into `Mono.defer()`

**Review by PavloEsco — DISMISSED** [review]
Tested OK

File: scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java (hunk 1 above)
Comment (stingion) [reply scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java:60]: Now it's done

File: scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java (hunk 2 above)
Comment (stingion) [reply scalecube-configuration/src/test/java/io/scalecube/configuration/repository/InMemoryConfigurationRepository.java:54]: It's local testing and performance is very high

File: scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java [hunk 4]
```diff
@@ -9,4 +10,15 @@
 
 @ExtendWith(Fixtures.class)
 @WithFixture(value = IntegrationEnvironmentFixture.class, lifecycle = Lifecycle.PER_METHOD)
-final class ReadListIT extends ReadListScenario {}
+final class ReadListIT extends ReadListScenario {
+
+  @Override
+  protected void readEntries(ConfigurationService configurationService) {
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java:16]: Why did you override it?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java [hunk 5]
```diff
@@ -9,4 +10,15 @@
...
+    // not used
+  }
+
+  @Override
+  protected void readEntriesWithSpecificVersions(ConfigurationService configurationService) {
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java:21]: Why did you override it?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/scenario/ReadListScenario.java [hunk 6]
```diff
@@ -14,248 +16,370 @@
...
+import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.TestTemplate;
 import org.testcontainers.shaded.org.apache.commons.lang.RandomStringUtils;
 import reactor.test.StepVerifier;
 
+/** Todo: need to be implemented in IT. If to stop a breakpoint on code it coming not empty. */
```
Comment (segabriel) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/scenario/ReadListScenario.java]: Does it still need?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java (hunk 4 above)
Comment (stingion) [reply scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java:16]: It's to long to test it should be to increase waiting for update from 1 to few seconds in each method and to create new env. So IT is going not a minute but 5-7 minutes. I'll explain you at office

File: scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java (hunk 5 above)
Comment (stingion) [reply scalecube-configuration/src/test/java/io/scalecube/configuration/it/ReadListIT.java:21]: Read above

**Review by segabriel — APPROVED** [review]
Anyway let QA approve these changes

**Review by PavloEsco — APPROVED** [review]
Tested Ok, benchmarks > https://scalecube-gateway-tests.s3.amazonaws.com/reports/1563788938488/ScalecubeConfigurationServiceBenchmark.zip

**Pull request merged**